src/compound.jl

```diff
@@ -0,0 +1,180 @@
+                                                                         Catalyst.CompoundComponents,
+                                                                         $species_expr))
+
+    # Ensure the expression is evaluated in the correct scope by escaping it
+    escaped_setcomponents_expr = esc(setcomponents_expr)
+
+    # Construct the expression to set the coefficients metadata
+    setcoefficients_expr = :($(species_name) = ModelingToolkit.setmetadata($(species_name),
+                                                                           Catalyst.CompoundCoefficients,
+                                                                           $coeffs_expr))
+
+    escaped_setcoefficients_expr = esc(setcoefficients_expr)
+
+    # Return a block that contains the escaped expressions
+    return Expr(:block, escaped_species_expr, escaped_setmetadata_expr,
+                escaped_setcomponents_expr, escaped_setcoefficients_expr)
+end
+
+
+# Check if a species is a compound
+iscompound(s::Num) = iscompound(MT.value(s))
+function iscompound(s)
+    MT.getmetadata(s, CompoundSpecies, false)
+end
+
+coefficients(s::Num) = coefficients(MT.value(s))
+function coefficients(s)
+    MT.getmetadata(s, CompoundCoefficients)
+end
+
+components(s::Num) = components(MT.value(s))
+function components(s)
+    MT.getmetadata(s, CompoundComponents)
+end
+
+component_coefficients(s::Num) = component_coefficients(MT.value(s))
+function component_coefficients(s)
+    return [c => co for (c, co) in zip(components(s), coefficients(s))]
+end
+
+### Balancing Code
+
+function create_matrix(reaction::Catalyst.Reaction)
+    compounds = [reaction.substrates; reaction.products]
+    atoms = []
+    n_atoms = 0
+    A = zeros(Int, 0, length(compounds))
+
+    for (j, compound) in enumerate(compounds)
+        if iscompound(compound)
+            pairs = component_coefficients(compound)
+            if pairs == Nothing 
+                continue
+            end
+        else 
+            pairs = [(compound, 1)]
+        end
+
+        for pair in pairs
+            atom, coeff = pair
+            i = findfirst(x -> isequal(x, atom), atoms)
+            if i === nothing  
+                push!(atoms, atom)
+                n_atoms += 1
+                A = [A; zeros(Int, 1, length(compounds))]
+                i = n_atoms
+            end
+            coeff = any(map(p -> isequal(p, compounds[j]), reaction.products)) ? -coeff : coeff
+            A[i, j] = coeff
+        end
+    end
+
+    # Append a row with last element as 1 
+    new_row = zeros(Int, 1, size(A, 2))
+    new_row[end] = 1
+    A = vcat(A, new_row)
+
+    return A
+end
+
+function get_stoich(reaction::Reaction)
+    # Create the matrix A using create_matrix function.
+    A = create_matrix(reaction)
+    
+    # Create the vector b. The last element is 1 and others are 0.
+    b = zeros(size(A,1))
+    b[end] = 1
+    
+    # Compute x= A^-1 *b
+    x = inv(A)*b
```

Comment:
You should never solve Ax = b via taking the inverse. You should use a linear system solver, in this case we probably want one that works in integer arithmetic to avoid rounding. Take a look at what we do to calculate nullspaces in Catalyst, and at the code in:

https://github.com/SciML/ModelingToolkit.jl/blob/master/src/structural_transformation/bareiss.jl

which is designed to do row reduction while preserving integer matrices.

Reply:
https://gregorygundersen.com/blog/2020/12/09/matrix-inversion/

Reply:
Understood! Thank you for the information.